File: java/src/main/java/com/epam/eng/pattern/strategy/ShoppingCart.java
package com.epam.eng.pattern.strategy;

import java.util.ArrayList;
import java.util.List;

public class ShoppingCart {
    private List<Item> items;

    public ShoppingCart() {
        items = new ArrayList<>();
    }

    public void addItem(Item item) {
        items.add(item);
    }

    public void removeItem(Item item) {
        items.remove(item);
    }

    public double calculateTotal() {
        double total = 0.0;
        for (Item item : items) {
            total += item.getPrice();
        }
        return total;
    }

    public void pay(String paymentMethod) {
        double total = calculateTotal();
        if (paymentMethod.equals("credit_card")) {
            System.out.println("Payment made with credit card: $" + total);
        } else if (paymentMethod.equals("paypal")) {
            System.out.println("Payment made with PayPal: $" + total);
        } else {
            throw new IllegalArgumentException("Unknown payment method: " + paymentMethod);
        }
    }

    class Item {
        private String name;
        private double price;

        public Item(String name, double price) {
            this.name = name;
            this.price = price;
        }

        public String getName() {
            return name;
        }

        public void setName(String name) {
            this.name = name;
        }

        public double getPrice() {
            return price;
        }

        public void setPrice(double price) {
            this.price = price;
        }
    }

}

// todo: consider to abstract payment

File: ts/strategy/ShoppingCart.ts
interface PaymentStrategy {
  pay(amount: number): void;
}

class CreditCardPayment implements PaymentStrategy {
  pay(amount: number): void {
    console.log("Payment made with credit card: $" + amount);
  }
}

class PayPalPayment implements PaymentStrategy {
  pay(amount: number): void {
    console.log("Payment made with PayPal: $" + amount);
  }
}

class ShoppingCart {
  private items: Item[] = [];

  constructor() {
    this.items = [];
  }

  public addItem(item: Item): void {
    this.items.push(item);
  }

  public removeItem(item: Item): void {
    this.items.splice(
      this.items.findIndex((i) => i === item),
      1
    );
  }

  public calculateTotal(): number {
    let total = 0.0;
    for (let item of this.items) {
      total += item.getPrice();
    }
    return total;
  }

  public pay(paymentStrategy: PaymentStrategy): void {
    const total = this.calculateTotal();
    paymentStrategy.pay(total);
  }
}

class Item {
  private name: string;
  private price: number;

  constructor(name: string, price: number) {
    this.name = name;
    this.price = price;
  }

  public getName(): string {
    return this.name;
  }

  public setName(name: string): void {
    this.name = name;
  }

  public getPrice(): number {
    return this.price;
  }

  public setPrice(price: number): void {
    this.price = price;
  }
}

const cart = new ShoppingCart();
const item1 = new Item("Book", 10);
const item2 = new Item("Pen", 2);

cart.addItem(item1);
cart.addItem(item2);

cart.pay(new CreditCardPayment());
cart.pay(new PayPalPayment());
